src/main/scala/org/broadinstitute/dsde/rawls/expressions/SlickExpressionParsing.scala

```diff
@@ -280,38 +307,111 @@ trait SlickExpressionParser extends JavaTokenParsers {
+      .andThen({ op(evaluator) })
+      .cleanUp({ _ => evaluator.dropExprEvalTempTable() })
   }
 
-  def evalFinalEntity(workspaceContext: SlickWorkspaceContext, rootType:String, rootName:String, expression:String): Try[ReadAction[Iterable[Entity]]] = {
-    parser.parseEntityExpr(expression)
-      .map( runPipe(SlickExpressionContext(workspaceContext, rootType, rootName), _).map(_.collect { case e: Entity => e } ))
+  def withNewExpressionEvaluator[R](parser: DataAccess, workspaceContext: SlickWorkspaceContext, rootType: String, rootName: String)
+                                   (op: SlickExpressionEvaluator => ReadWriteAction[R])
+                                   (implicit executionContext: ExecutionContext): ReadWriteAction[R] = {
+    import parser.driver.api._
+
+    //Find the root entity for the expression
+    val dbRootEntityRec = parser.entityQuery.findEntityByName(workspaceContext.workspaceId, rootType, rootName).result
+
+    //Sanity check that we've only got one, and then pass upwards
+    dbRootEntityRec flatMap { rootEntityRec =>
+      if(rootEntityRec.size != 1) {
+        DBIO.failed(new RawlsException(s"Found != 1 root entity when searching for ${rootType}/$rootName"))
+      } else {
+        withNewExpressionEvaluator(parser, rootEntityRec)(op)
+      }
+    }
+  }
+}
+
+class SlickExpressionEvaluator(val parser: DataAccess, val rootEntities: Seq[EntityRecord])(implicit executionContext: ExecutionContext) {
+  import parser.driver.api._
+
+  def populateExprEvalTempTable() = {
+    parser.exprEvalQuery ++= rootEntities.map( e => parser.ExprEvalRecord(e.name) )
+  }
+
+  def dropExprEvalTempTable() = {
+    parser.exprEvalQuery.delete
+  }
+
+  def evalFinalAttribute(workspaceContext: SlickWorkspaceContext, expression: String): ReadWriteAction[Map[String, Try[Iterable[AttributeValue]]]] = {
+    parser.parseAttributeExpr(expression) match {
+      case Failure(regret) => DBIO.failed(new RawlsException(regret.getMessage))
+      case Success(expr) =>
+        runPipe(SlickExpressionContext(workspaceContext, rootEntities), expr) map { exprResults =>
+          val results = exprResults map { case (key, attrVals) =>
+            key -> Try(attrVals.collect {
+              case AttributeNull => Seq.empty
+              case AttributeEmptyList => Seq.empty
+              case av: AttributeValue => Seq(av)
+              case avl: AttributeValueList => avl.list
+              case badType => throw new RawlsException(s"unsupported type resulting from attribute expression: $badType: ${badType.getClass}")
+            }.flatten)
+          }
+          //add any missing entities (i.e. those missing the attribute) back into the result map
+          results ++ rootEntities.map(_.name).filterNot( results.keySet.contains ).map { missingKey => missingKey -> Success(Seq()) }
+        }
+    }
+  }
+
+  //This is boiling away the Try associated with attempting to parse the expression. Is this OK?
```

Comment:
Oh! I guess. But it couples evaluating entities to evaluating attributes somewhat. I mean, we could go all the way and make the call "evaluate this attribute expression on the results of this set of entity expressions" but that strikes me as doing rather a lot in one place. IMHO we should maintain the distinction.